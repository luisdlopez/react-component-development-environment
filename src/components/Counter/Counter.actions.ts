export const INCREMENT = 'Counter/INCREMENT';
export const DECREMENT = 'Counter/DECREMENT';

export interface IActionSignature {
  readonly type: string;
}

export function increment(): IActionSignature {
  return {
    type: INCREMENT,
  };
}

export function decrement(): IActionSignature {
  return {
    type: DECREMENT,
  };
}
