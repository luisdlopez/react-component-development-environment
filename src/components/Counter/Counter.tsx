import * as React from 'react';

export interface ICounterProps {
  readonly value?: number;
  readonly onIncrement?: () => null;
  readonly onDecrement?: () => null;
}

export default class Counter extends React.Component<ICounterProps> {
  constructor(props: ICounterProps) {
    super(props);
    this.increment = this.increment.bind(this);
    this.decrement = this.decrement.bind(this);
  }

  public render(): JSX.Element {
    const { value } = this.props;
    return (
      <div>
        <p>Current counter value: <span>{value}</span></p>
        <div>
          <button onClick={this.increment}>Increment</button>
          <button onClick={this.decrement}>Decrement</button>
        </div>
      </div>
    );
  }

  private increment(): void {
    if (this.props.onIncrement) {
      this.props.onIncrement();
    }
  }

  private decrement(): void {
    if (this.props.onDecrement) {
      this.props.onDecrement();
    }
  }
}
