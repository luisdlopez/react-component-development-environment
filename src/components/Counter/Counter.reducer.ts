import { DECREMENT, INCREMENT } from './Counter.actions';

const DEFAULT_STATE = 0;

export default function reducer(state?: number, action?: any): number {
  if (action) {
    switch (action.type) {
      case INCREMENT:
        return state + 1;
      case DECREMENT:
        return state - 1;
      default:
        return state ? state : DEFAULT_STATE;
    }
  } else {
    return state ? state : DEFAULT_STATE;
  }
}
