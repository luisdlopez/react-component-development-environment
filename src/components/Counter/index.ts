import Counter from './Counter';
import { decrement, increment, IActionSignature } from './Counter.actions';
import reducer from './Counter.reducer';

const CounterWrapper = {
  Counter,
  reducer,
  actions: {
    decrement,
    increment,
  },
};

export { CounterWrapper };
export default CounterWrapper;
