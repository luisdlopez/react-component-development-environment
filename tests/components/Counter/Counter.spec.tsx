import * as React from 'react';
import Counter, { ICounterProps } from '../../../src/components/Counter/Counter';
import { shallow } from 'enzyme';

describe('Counter', () => {

  const DEFAULT_PROPS: ICounterProps = {
    value: 0,
    onDecrement: () => null,
    onIncrement: () => null,
  };

  test('basic snapshot', () => {
    const wrapper = shallow(<Counter {...DEFAULT_PROPS}/>);
    expect(wrapper).toMatchSnapshot();
  });

  describe('value displayed', () => {

    test('value displayed should be 0 when value passed as props is 0', () => {
      const wrapper = shallow(<Counter {...DEFAULT_PROPS}/>);
      // text inside the span that displayed the current counter value
      const text = wrapper.find('p > span').text();

      expect(text).toBe('0');
    });

    test('value displayed should be -1 when value passed as props is -1', () => {
      const props = Object.assign({}, DEFAULT_PROPS, { value: -1 });
      const wrapper = shallow(<Counter {...props}/>);
      // text inside the span that displayed the current counter value
      const text = wrapper.find('p > span').text();

      expect(text).toBe('-1');
    });

    test('value displayed should be 1 when value passed as props is 1', () => {
      const props = Object.assign({}, DEFAULT_PROPS, { value: 1 });
      const wrapper = shallow(<Counter {...props}/>);
      // text inside the span that displayed the current counter value
      const text = wrapper.find('p > span').text();

      expect(text).toBe('1');
    });

  });

  describe('actions', () => {

    test('onIncrement should be called when button "Increment" is clicked', () => {
      const props = Object.assign({}, DEFAULT_PROPS, { onIncrement: jest.fn() });
      const wrapper = shallow(<Counter {...props}/>);
      // click first button (increment)
      wrapper.find('button').at(0).simulate('click');

      expect(props.onIncrement).toHaveBeenCalledTimes(1);
    });

    test('onDecrement should be called when button "Decrement" is clicked', () => {
      const props = Object.assign({}, DEFAULT_PROPS, { onDecrement: jest.fn() });
      const wrapper = shallow(<Counter {...props}/>);
      // click second button (decrement)
      wrapper.find('button').at(1).simulate('click');

      expect(props.onDecrement).toHaveBeenCalledTimes(1);
    });

  });

});
