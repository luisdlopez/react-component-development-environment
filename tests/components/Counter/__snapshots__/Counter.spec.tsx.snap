// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter basic snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Counter
    onDecrement={[Function]}
    onIncrement={[Function]}
    value={0}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <p>
          Current counter value: 
          <span>
            0
          </span>
        </p>,
        <div>
          <button
            onClick={[Function]}
          >
            Increment
          </button>
          <button
            onClick={[Function]}
          >
            Decrement
          </button>
        </div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Current counter value: ",
            <span>
              0
            </span>,
          ],
        },
        "ref": null,
        "rendered": Array [
          "Current counter value: ",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": 0,
            },
            "ref": null,
            "rendered": 0,
            "type": "span",
          },
        ],
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <button
              onClick={[Function]}
            >
              Increment
            </button>,
            <button
              onClick={[Function]}
            >
              Decrement
            </button>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Increment",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Increment",
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Decrement",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Decrement",
            "type": "button",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <p>
            Current counter value: 
            <span>
              0
            </span>
          </p>,
          <div>
            <button
              onClick={[Function]}
            >
              Increment
            </button>
            <button
              onClick={[Function]}
            >
              Decrement
            </button>
          </div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Current counter value: ",
              <span>
                0
              </span>,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Current counter value: ",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": 0,
              },
              "ref": null,
              "rendered": 0,
              "type": "span",
            },
          ],
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <button
                onClick={[Function]}
              >
                Increment
              </button>,
              <button
                onClick={[Function]}
              >
                Decrement
              </button>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Increment",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Increment",
              "type": "button",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Decrement",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Decrement",
              "type": "button",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
